### PR TITLE
feat(map): add clustering and pin bitmap generation utilities

### LINKED ISSUE #512: Implement Clustering & Pin Generation Logic
### Description
Implement the core mathematical utilities for map event clustering and cluster pin generation.

### Acceptance Criteria
- [ ] Create MapClustering utility with Union-Find algorithm for distance-based clustering
- [ ] Implement Haversine distance calculation for geographic clustering
- [ ] Create PinBitmapGenerator for dynamic cluster pin generation
- [ ] Handle edge cases (same location events, large clusters >99)
- [ ] Provide 100% unit test coverage for clustering logic
- [ ] Ensure cluster pins use correct theme colors and scaling

### PR DESCRIPTION
**Pull request by M-e-d-A**
# Description:
## Purpose of the change
This PR introduces the core utility classes for map event clustering, as defined in sub-issue #512.

It includes:
* `MapClustering` utility with Union-Find algorithm and Haversine distance calculation for grouping nearby events
* `PinBitmapGenerator` for creating cluster annotation images with event counts
* Comprehensive unit tests for both clustering logic and bitmap generation
* Support for same-location event stacking and zoom-based clustering

## Related issues
Closes #512 
Part of #511 

## How to test
1. Check out the branch.
2. Run the tests in `MapClusteringTest.kt` and `PinBitmapGeneratorTest.kt` via Android Studio or `./gradlew test`.
3. Verify all tests pass.
4. Review the clustering algorithm and bitmap generation logic for correctness.

**Note:** An LLM was used for the sole and exclusive purpose of improving the clarity and formatting of the PR description.

**Related PRs in this series:**
- **#514** *(this PR)* 
- #515

**Screenshots of the final product:**

<table>
  <tr>
    <td align="center">
      Four events on the exact same location A
      <br>
      <em>Four events on the exact same location A</em>
    </td>
    <td align="center">
      Map with clustering A
      <br>
      <em>Map with clustering A</em>
    </td>
  </tr>
  <tr>
    <td align="center">
      Map with clustering B
      <br>
      <em>Map with clustering B</em>
    </td>
    <td align="center">
      Four events on the exact same location B
      <br>
      <em>Four events on the exact same location B</em>
    </td>
  </tr>
</table>

### CODE REVIEW

File: app/src/main/java/ch/onepass/onepass/ui/map/MapClustering.kt
Comment (RD418) [inline app/src/main/java/ch/onepass/onepass/ui/map/MapClustering.kt]: Great code ! Really, your code is clean, well-documented. And super idea to use union-find for clusters.

File: app/src/test/java/ch/onepass/onepass/ui/map/MapClusteringTest.kt
Comment (RD418) [inline app/src/test/java/ch/onepass/onepass/ui/map/MapClusteringTest.kt]: Your tests looks ok to me great job !  Your line coverage is 98% !

**Review by RD418 — CHANGES_REQUESTED** [review]
### Overall
Your code is super clear and documented. Excellent code coverage, but I have to request change on your forgotten hard-coded color, please check below !

File: app/src/main/java/ch/onepass/onepass/ui/map/MapClustering.kt
Comment (M-e-d-A) [reply app/src/main/java/ch/onepass/onepass/ui/map/MapClustering.kt]: Thank you!

File: app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    val bitmap = Bitmap.createBitmap(size.toInt(), size.toInt(), Bitmap.Config.ARGB_8888)
+    val canvas = Canvas(bitmap)
+    val paint = Paint(Paint.ANTI_ALIAS_FLAG)
+
+    val cx = size / 2f
+    val cy = size / 2f
+    val radius = size / 2.5f
+
+    // Draw Circle Background (OnePass Purple)
+    paint.color = Color.parseColor("#6200EE") // Replace with your theme color
```
[the remark this answers is not in the document]
Comment (M-e-d-A) [reply app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt]: Thanks for spotting it, fixed!

File: app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    paint.style = Paint.Style.STROKE
+    paint.strokeWidth = 4f * scale
+    canvas.drawCircle(cx, cy, radius, paint)
+
+    // Draw Text
+    paint.style = Paint.Style.FILL
+    paint.color = Color.WHITE
+    paint.textAlign = Paint.Align.CENTER
+    paint.textSize = 24f * scale
+    paint.typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
```
[the remark this answers is not in the document]
Comment (M-e-d-A) [reply app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt]: Thanks for the remark! Resolved!

File: app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt
Comment (RD418) [inline app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt]: You fixed the requested changes ! GJ.

**Review by RD418 — APPROVED** [review]
### Update after RC
Everything is fixed, good job.

File: app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt
Comment (M-e-d-A) [reply app/src/main/java/ch/onepass/onepass/ui/map/PinBitmapGenerator.kt]: Thanks!
